### PR TITLE
DPIB/posbuf pointer mode enabling

### PR DESCRIPTION
**Pull request by keyonjie**
This series is to enable DPIB/posbuf pointer mode for SOF on SKL+ platforms:

   1. refine BDL settings.
    2. add pointer callback to read postion from DPIB/posbuf;
    3. add Kconfig to force IPC update mode on SKL+, that is:
    for SKL-, we will always use IPC update mode;
    for SKL+, we will use DPIB/posbuf by default, but be able to force IPC mode for debug.

Hi, @keqiaozhang @ZhendanYang, can you help verified if this will cause any regression on existed(e.g. byt and apl)?

Hi, @lgirdwood please only merge after verification passed.

### CODE REVIEW

File: sound/soc/sof/intel/hda-pcm.c [hunk 1]
```diff
@@ -136,6 +137,38 @@ int hda_dsp_pcm_trigger(struct snd_sof_dev *sdev,
...
 
+snd_pcm_uframes_t hda_dsp_pcm_pointer(struct snd_sof_dev *sdev,
+				      struct snd_pcm_substream *substream)
+{
+	struct hdac_stream *hstream = substream->runtime->private_data;
+	struct hdac_ext_stream *stream = stream_to_hdac_ext_stream(hstream);
+	struct hdac_bus *bus = sof_to_bus(sdev);
+	snd_pcm_uframes_t pos = 0;
+
+	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK) {
```
Comment (lgirdwood) [inline sound/soc/sof/intel/hda-pcm.c]: please comment on logic here and why there is difference for playback and capture.

File: sound/soc/sof/intel/hda-pcm.c [hunk 2]
```diff
@@ -136,6 +137,38 @@ int hda_dsp_pcm_trigger(struct snd_sof_dev *sdev,
...
+		pos = snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR,
+				       AZX_REG_VS_SDXDPIB_XBASE +
+				       (AZX_REG_VS_SDXDPIB_XINTERVAL *
+					hstream->index));
+	} else {
+		usleep_range(20, 30);
```
Comment (lgirdwood) [inline sound/soc/sof/intel/hda-pcm.c]: why this time, pls comment

File: sound/soc/sof/intel/hda-pcm.c [hunk 3]
```diff
@@ -136,6 +137,38 @@ int hda_dsp_pcm_trigger(struct snd_sof_dev *sdev,
...
+				  hstream->index));
+		pos = snd_hdac_stream_get_pos_posbuf(hstream);
+	}
+
+	if (pos >= hstream->bufsize)
+		pos = 0;
+
+	pos = bytes_to_frames(substream->runtime, pos);
+
+	dev_dbg(sdev->dev, "PCM: stream %d dir %d position %lu\n",
```
Comment (lgirdwood) [inline sound/soc/sof/intel/hda-pcm.c]: dev_vdbg since this will be high frequency ?

File: sound/soc/sof/ipc.c [hunk 4]
```diff
@@ -591,7 +591,14 @@ static void ipc_stream_message(struct snd_sof_dev *sdev, u32 msg_cmd)
 
 	switch (msg_type) {
 	case SOF_IPC_STREAM_POSITION:
-		ipc_period_elapsed(sdev, msg_id);
+		if (!sdev->ops->pcm_pointer ||
+		    IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_FORCE_IPC_POSITION)) {
```
Comment (lgirdwood) [inline sound/soc/sof/ipc.c]: can you make this #if like the others for consistency

File: sound/soc/sof/intel/hda-stream.c [hunk 5]
```diff
@@ -36,57 +36,97 @@
 #include "hda.h"
 
 /*
- * set up Buffer Descriptor List (BDL) for host memory transfer
- * BDL describes the location of the individual buffers and is little endian.
+ * set up one of BDL entries for a stream
```
Comment (lgirdwood) [inline sound/soc/sof/intel/hda-stream.c:39]: Why the change, there is no explanation in the commit message. What is wrong with existing BDL func ?

File: sound/soc/sof/ipc.c (hunk 4 above)
Comment (ranj063) [reply sound/soc/sof/ipc.c]: @keyonjie is this line correct, should this be !IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_FORCE_IPC_POSITION)?

**Comment by ranj063** [comment]
@keyonjie recently while trying to no_irq mode in the FW, I ran into trouble with the playback not being able to continue without the perio update ipc from the FW.

Will these commit help with that?

File: sound/soc/sof/ipc.c (hunk 4 above)
Comment (keyonjie) [reply sound/soc/sof/ipc.c]: @ranj063 we handle ipc and run ipc_elapse here when force_ipc_position is configured, or we don't have pcm_pointer callback, so this line is correct.

Comment (ranj063) [reply sound/soc/sof/ipc.c]: @keyonjie makes sense. I misunderstood the config to be to force no ipc update. Thanks!

**Comment by ranj063** [comment]
@keyonjie if we use psbuf by default, should the no_irq mode be enabled  by default too?

File: sound/soc/sof/intel/hda-pcm.c (hunk 2 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-pcm.c]: OK let me add comments to this logic.

File: sound/soc/sof/intel/hda-pcm.c (hunk 3 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-pcm.c]: good, let me change.

File: sound/soc/sof/ipc.c (hunk 4 above)
Comment (keyonjie) [reply sound/soc/sof/ipc.c]: @lgirdwood , if use #if IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_FORCE_IPC_POSITION) here, the logic may looks more ugly here?

File: sound/soc/sof/intel/hda-stream.c (hunk 5 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-stream.c:39]: in the existing BDL, we don't support interrupt mode with DIPB/posbuf, let me add commit message to explain it.

**Comment by keyonjie** [comment]
@ranj063 how did you try no_irq mode?

**Comment by ranj063** [comment]
@keyonjie setting host_period_bytes in stream params to 0 enables no_irq mode in the FW.

**Comment by lgirdwood** [comment]
@ranj063 no_irq mode works by userspace asking driver for position updates. driver then has to read from SRAM window or DPIB. The playback may stop if the position does not move enough in the given time period.

**Comment by ranj063** [comment]
@lgirdwood, does the DPIB posbuf update apply to normal SSP playback as well or is it meant for HDMI?

**Comment by ranj063** [comment]
@keyonjie I tried your building your patchset for glk and saw a couple of compilation errors related to unused variables in hda-stream.c lines 99 and 102 (addr and pos_align)

**Comment by ranj063** [comment]
@keyonjie this patchset breaks HDMI playback on glk

**Comment by keyonjie** [comment]
@ranj063 thanks for your feedback, actually it's fixed locally but looks update not pushed yet, doing that now.

**Comment by keyonjie** [comment]
@ranj063 can you share how it breaking HDMI?

**Comment by keyonjie** [comment]
@lgirdwood can you share how to configure to use no_irq(SNDRV_PCM_HW_PARAMS_NO_PERIOD_WAKEUP) when run aplay?

**Comment by ranj063** [comment]
@keyonjie I get the "Unable to install hw params" error with aplay -D hw:0,5 for HDMI playback on the Yorp

**Comment by keyonjie** [comment]
@lgirdwood @ranj063 all comments are addressed, I tried on APL, both set/unset FORCE_IPC_POSITION, the position update for HDMI works on my side.

**Comment by keyonjie** [comment]
@keqiaozhang @ZhendanYang @xiulipan or anybody who can you help with a quick try on your byt/cnl SSP/DMIC to check if there is any regression with it?

**Comment by keqiaozhang** [comment]
@keyonjie @lgirdwood 
There's no regression on BYT. But on APL, there is a obvious noise, checked with trace log and dmesg, no error message output.
I also tried to enable/disable CONFIG_SND_SOC_SOF_DEBUG_FORCE_IPC_POSITION  in kernel config, found that when enabling CONFIG_SND_SOC_SOF_DEBUG_FORCE_IPC_POSITION, the noise is worse than disabling it.

**Comment by keyonjie** [comment]
thanks @keqiaozhang , it's not too bad to me actually. Let me debug to root cause why it introduce noise on APL.

**Comment by keyonjie** [comment]
@lgirdwood @keqiaozhang just root caused and fixed the noise issue. 
I just verified it works fine for both DPIB/IPC position mode, with SSP/HDMI/DP, on BYT/APL/CNL.

I think it's ready for merge now.

**Comment by keyonjie** [comment]
@ranj063 please use this latest version for your HDMI related works also.
